frontend/src/Context/BackgroundPublisherProvider/useBackgroundPublisher/useBackgroundPublisher.tsx

```diff
@@ -0,0 +1,234 @@
+    (backgroundSelected: string) => {
+      applyBackgroundFilter(
+        backgroundPublisherRef.current,
+        backgroundSelected,
+        undefined,
+        setBackgroundFilter,
+        false
+      ).catch(() => {
+        console.error('Failed to apply background filter.');
+      });
+    },
+    [setBackgroundFilter]
+  );
+
+  /**
+   * Change video camera in use
+   * @returns {void}
+   */
+  const changeVideoSource = useCallback((deviceId: string) => {
+    if (!deviceId || !backgroundPublisherRef.current) {
+      return;
+    }
+    backgroundPublisherRef.current.setVideoSource(deviceId);
+    setLocalVideoSource(deviceId);
+  }, []);
+
+  /**
+   * Handle device permissions denial
+   * used to inform the user they need to give permissions to devices to access the call
+   * after a user grants permissions to the denied device, trigger a reload.
+   * @returns {void}
+   */
+  const handleBackgroundAccessDenied = useCallback(
+    async (event: AccessDeniedEvent) => {
+      const deviceDeniedAccess = event.message?.startsWith('Microphone') ? 'microphone' : 'camera';
+
+      setAccessStatus(DEVICE_ACCESS_STATUS.REJECTED);
+
+      try {
+        const permissionStatus = await window.navigator.permissions.query({
+          name: deviceDeniedAccess,
+        });
+        permissionStatus.onchange = () => {
+          if (permissionStatus.state === 'granted') {
+            setAccessStatus(DEVICE_ACCESS_STATUS.ACCESS_CHANGED);
+          }
+        };
+      } catch (error) {
+        console.error(`Failed to query device permission for ${deviceDeniedAccess}: ${error}`);
+      }
+    },
+    [setAccessStatus]
+  );
+
+  const handleVideoElementCreated = (event: PublisherVideoElementCreatedEvent) => {
+    setPublisherVideoElement(event.element);
+    setIsPublishing(true);
+  };
+
+  const addPublisherListeners = useCallback(
+    (publisher: Publisher | null) => {
+      if (!publisher) {
+        return;
+      }
+      publisher.on('destroyed', handleBackgroundDestroyed);
+      publisher.on('accessDenied', handleBackgroundAccessDenied);
+      publisher.on('videoElementCreated', handleVideoElementCreated);
+      publisher.on('accessAllowed', () => {
+        setAccessStatus(DEVICE_ACCESS_STATUS.ACCEPTED);
+        getAllMediaDevices();
+      });
+    },
+    [getAllMediaDevices, handleBackgroundAccessDenied, setAccessStatus]
+  );
+
+  const initBackgroundLocalPublisher = useCallback(async () => {
+    if (backgroundPublisherRef.current) {
+      return;
+    }
+
+    // Set videoFilter based on user's selected background
+    let videoFilter: VideoFilter | undefined;
+    if (initialBackgroundRef.current && hasMediaProcessorSupport()) {
+      videoFilter = initialBackgroundRef.current;
+    }
+
+    await deviceStoreRef.current.init();
+    const videoSource = deviceStoreRef.current.getConnectedDeviceId('videoinput');
+    const audioSource = deviceStoreRef.current.getConnectedDeviceId('audioinput');
```

Comment:
I'm curious if you could remove the logic around `audioSource` from this copy-pasta'ed part. I don't think it's needed in your use case